### PR TITLE
Add HTTP Proxy support

### PR DESCRIPTION
**Pull request by VitaliiTytarenko-okta**
OKTA-336114

### CODE REVIEW

**Review by bdemers — COMMENTED** [review]
IMHO, both the “standard” Java proxy props should work, and the ones used by the Okta SDK.

that way a user only has to specify the props once 🤔

**Comment by arvindkrishnakumar-okta** [comment]
> IMHO, both the “standard” Java proxy props should work, and the ones used by the Okta SDK.
> 
> that way a user only has to specify the props once 🤔

You might want to check out https://github.com/okta/okta-spring-boot#proxy to see if it helps.

**Comment by VitaliiTytarenko-okta** [comment]
> IMHO, both the “standard” Java proxy props should work, and the ones used by the Okta SDK.
> 
> that way a user only has to specify the props once 🤔

As I mentioned here #57, I think, it's a bad idea to mix these properties.

**Comment by bdemers** [comment]
We should standardize on `okta.oauth2.proxy.*` That makes it mirror what is used for the SDKs

Re: the property names, other libraries use the Java proxy properties or directly use the ProxySelector (IIRC OK HTTP uses the ProxySelector).

**Comment by VitaliiTytarenko-okta** [comment]
@bdemers 
Could you take a look, please?
I changed property names. Have I missed something else?

File: oauth2/pom.xml [hunk 1]
```diff
@@ -110,6 +110,11 @@
             <version>2.25.0</version>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>org.mock-server</groupId>
+            <artifactId>mockserver-netty</artifactId>
+            <version>5.11.2</version>
```
Comment (bdemers) [inline oauth2/pom.xml:116]: mark this as test scope

File: oauth2/src/main/java/com/okta/spring/boot/oauth/OktaOAuth2AutoConfig.java [hunk 2]
```diff
@@ -70,6 +82,52 @@ OidcClientInitiatedLogoutSuccessHandler oidcLogoutSuccessHandler(OktaOAuth2Prope
...
+
+        OktaOAuth2Properties.Proxy proxyProperties = oktaOAuth2Properties.getProxy();
+        if (proxyProperties != null && Strings.hasText(proxyProperties.getHost()) && proxyProperties.getPort() != 0) {
+            proxy = new Proxy(Proxy.Type.HTTP, new InetSocketAddress(proxyProperties.getHost(), proxyProperties.getPort()));
+
+            if (Strings.hasText(proxyProperties.getUsername()) &&
+                Strings.hasText(proxyProperties.getPassword())) {
+
+                Authenticator.setDefault(new ProxyPasswordAuthentication(proxyProperties.getUsername(),
+                    proxyProperties.getPassword().toCharArray()));
```
Comment (bdemers) [inline oauth2/src/main/java/com/okta/spring/boot/oauth/OktaOAuth2AutoConfig.java:99]: Don't set the default authenticator, this needs to be scoped to this RestTemplate instance

File: oauth2/src/main/java/com/okta/spring/boot/oauth/config/OktaOAuth2Properties.java [hunk 3]
```diff
@@ -179,4 +192,47 @@ public void validate(Object target, Errors errors) {
...
+            this.host = host;
+        }
+
+        public int getPort() {
+            return port;
+        }
+
+        public void setPort(int port) {
+            this.port = port;
+        }
```
Comment (bdemers) [inline oauth2/src/main/java/com/okta/spring/boot/oauth/config/OktaOAuth2Properties.java:220]: nit: consider validating the `Proxy` object above you have an if statement that both `host` and `port` either both required or both should be missing, but there is no validation, if one of them is missing

File: oauth2/src/test/groovy/com/okta/spring/boot/oauth/OktaOAuth2AutoConfigRestTemplateTest.groovy [hunk 4]
```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright 2020-Present Okta, Inc.
```
Comment (bdemers) [inline oauth2/src/test/groovy/com/okta/spring/boot/oauth/OktaOAuth2AutoConfigRestTemplateTest.groovy:2]:
```suggestion
 * Copyright 2021-Present Okta, Inc.
```
Check other files added in this commit

File: oauth2/src/test/groovy/com/okta/spring/boot/oauth/OktaOAuth2AutoConfigRestTemplateTest.groovy [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+        OktaOAuth2Properties properties = new OktaOAuth2Properties(null)
+        properties.setProxy(proxy)
+
+        RestTemplate restTemplate = new OktaOAuth2AutoConfig().restTemplate(properties)
+        def headers = new HttpHeaders(singletonMap("Cookie", "sessionId=" + sessionId))
+        ResponseEntity<OAuth2AccessTokenResponse> response = restTemplate
+            .exchange("http://base_url.com", HttpMethod.GET, new HttpEntity<String>(headers), OAuth2AccessTokenResponse)
+        verify(response, sessionId)
+        assertThat "Wrong user", Authenticator.theAuthenticator.getAt("proxyUser"), is("foo")
+        assertThat "Wrong password", Authenticator.theAuthenticator.getAt("proxyPassword").toString(), is("bar")
```
Comment (bdemers) [inline oauth2/src/test/groovy/com/okta/spring/boot/oauth/OktaOAuth2AutoConfigRestTemplateTest.groovy:78]: nit: add the word `proxy` here

File: oauth2/src/test/groovy/com/okta/spring/boot/oauth/OktaOAuth2AutoConfigRestTemplateTest.groovy [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+import org.testng.annotations.Test
+
+import static java.util.Collections.singletonList
+import static java.util.Collections.singletonMap
+import static org.hamcrest.MatcherAssert.assertThat
+import static org.hamcrest.Matchers.is
+
+class OktaOAuth2AutoConfigRestTemplateTest {
+    private static final String LOCALHOST = "localhost"
+    private static final int PORT = 7000
```
Comment (bdemers) [inline oauth2/src/test/groovy/com/okta/spring/boot/oauth/OktaOAuth2AutoConfigRestTemplateTest.groovy:42]: Don't use a static port for testing, use a random free port, this mockserver _should_ support this out of the box, but if not, just lookup a free port first

File: oauth2/src/test/groovy/com/okta/spring/boot/oauth/OktaOAuth2AutoConfigRestTemplateTest.groovy [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+        proxy.setPort(PORT)
+        proxy.setUsername("foo")
+        proxy.setPassword("bar")
+        OktaOAuth2Properties properties = new OktaOAuth2Properties(null)
+        properties.setProxy(proxy)
+
+        RestTemplate restTemplate = new OktaOAuth2AutoConfig().restTemplate(properties)
+        def headers = new HttpHeaders(singletonMap("Cookie", "sessionId=" + sessionId))
+        ResponseEntity<OAuth2AccessTokenResponse> response = restTemplate
+            .exchange("http://base_url.com", HttpMethod.GET, new HttpEntity<String>(headers), OAuth2AccessTokenResponse)
```
Comment (bdemers) [inline oauth2/src/test/groovy/com/okta/spring/boot/oauth/OktaOAuth2AutoConfigRestTemplateTest.groovy:75]: use `example.com`

**Review by bdemers — CHANGES_REQUESTED** [review]

File: README.md [hunk 8]
```diff
@@ -361,20 +361,29 @@ This module integrates with Spring Security's OAuth support, all you need is the
...
+```
+
+or, add JVM args to your application like:
 
 ```bash
--Dhttps.proxyHost=host
--Dhttps.proxyPort=port
--Dhttps.proxyUser="user"          # optional
--Dhttps.proxyPassword="password". # optional 
+-Dokta.oauth2.proxy.host=host
```
Comment (arvindkrishnakumar-okta) [inline README.md:378]: Please add `username` and `password` properties here.

File: README.md [hunk 9]
```diff
@@ -361,20 +361,29 @@ This module integrates with Spring Security's OAuth support, all you need is the
...
+-Dokta.oauth2.proxy.port=port
 ```
 
 or, you could set it programmatically like:
 
 ```java
-System.setProperty("https.proxyHost", "https://example-proxy.com");
-System.setProperty("https.proxyPort", "443");
+System.setProperty("okta.oauth2.proxy.host", "example-proxy.com");
```
Comment (arvindkrishnakumar-okta) [inline README.md:385]: same as above comment.

File: oauth2/pom.xml (hunk 1 above)
Comment (VitaliiTytarenko-okta) [reply oauth2/pom.xml:116]: 🆗

File: oauth2/src/main/java/com/okta/spring/boot/oauth/OktaOAuth2AutoConfig.java (hunk 2 above)
Comment (VitaliiTytarenko-okta) [reply oauth2/src/main/java/com/okta/spring/boot/oauth/OktaOAuth2AutoConfig.java]: 👍

File: oauth2/src/test/groovy/com/okta/spring/boot/oauth/OktaOAuth2AutoConfigRestTemplateTest.groovy (hunk 4 above)
Comment (VitaliiTytarenko-okta) [reply oauth2/src/test/groovy/com/okta/spring/boot/oauth/OktaOAuth2AutoConfigRestTemplateTest.groovy]: 👌

File: oauth2/src/test/groovy/com/okta/spring/boot/oauth/OktaOAuth2AutoConfigRestTemplateTest.groovy (hunk 6 above)
Comment (VitaliiTytarenko-okta) [reply oauth2/src/test/groovy/com/okta/spring/boot/oauth/OktaOAuth2AutoConfigRestTemplateTest.groovy]: 👍

File: oauth2/src/test/groovy/com/okta/spring/boot/oauth/OktaOAuth2AutoConfigRestTemplateTest.groovy (hunk 7 above)
Comment (VitaliiTytarenko-okta) [reply oauth2/src/test/groovy/com/okta/spring/boot/oauth/OktaOAuth2AutoConfigRestTemplateTest.groovy]: 🆗

File: README.md (hunk 9 above)
Comment (VitaliiTytarenko-okta) [reply README.md:387]: 🆗

File: README.md (hunk 8 above)
Comment (VitaliiTytarenko-okta) [reply README.md:378]: 🆗

File: README.md [hunk 10]
```diff
@@ -361,20 +361,33 @@ This module integrates with Spring Security's OAuth support, all you need is the
 
 ## Proxy
 
-If you're running your application (with this okta-spring-boot dependency) from behind a network proxy, you could add JVM args to your application like:
+If you're running your application (with this okta-spring-boot dependency) from behind a network proxy, you could setup properties for it in application.yml:
+```yaml
+okta:
+  oauth2:
+    proxy:
+      host: "example-proxy.com"
```
Comment (bdemers) [inline README.md:369]: use `proxy.example.com` (when possible use `example.com` as this is the intended use for this domain)

File: README.md [hunk 11]
```diff
@@ -361,20 +361,33 @@ This module integrates with Spring Security's OAuth support, all you need is the
...
+-Dokta.oauth2.proxy.username=username
+-Dokta.oauth2.proxy.password=password
 ```
 
 or, you could set it programmatically like:
 
 ```java
-System.setProperty("https.proxyHost", "https://example-proxy.com");
-System.setProperty("https.proxyPort", "443");
+System.setProperty("okta.oauth2.proxy.host", "example-proxy.com");
```
Comment (bdemers) [inline README.md:387]:
```suggestion
System.setProperty("okta.oauth2.proxy.host", "proxy.example.com");
```

File: README.md [hunk 12]
```diff
@@ -361,20 +361,33 @@ This module integrates with Spring Security's OAuth support, all you need is the
...
 
 ```bash
--Dhttps.proxyHost=host
--Dhttps.proxyPort=port
--Dhttps.proxyUser="user"          # optional
--Dhttps.proxyPassword="password". # optional 
+-Dokta.oauth2.proxy.host=proxy.example.com
+-Dokta.oauth2.proxy.port=port
+-Dokta.oauth2.proxy.username=username
+-Dokta.oauth2.proxy.password=password
```
Comment (bdemers) [inline README.md:381]:
```suggestion
-Dokta.oauth2.proxy.username=your-username            # optional
-Dokta.oauth2.proxy.password=your-secret-password # optional
```

File: README.md [hunk 13]
```diff
@@ -361,20 +361,33 @@ This module integrates with Spring Security's OAuth support, all you need is the
...
 
 or, you could set it programmatically like:
 
 ```java
-System.setProperty("https.proxyHost", "https://example-proxy.com");
-System.setProperty("https.proxyPort", "443");
+System.setProperty("okta.oauth2.proxy.host", "proxy.example.com");
+System.setProperty("okta.oauth2.proxy.port", "7000");
+System.setProperty("okta.oauth2.proxy.username", "username");
+System.setProperty("okta.oauth2.proxy.password", "password");
```
Comment (bdemers) [inline README.md:390]: see above suggestion

**Comment by VitaliiTytarenko-okta** [comment]
Done

**Comment by VitaliiTytarenko-okta** [comment]
@bdemers @arvindkrishnakumar-okta
Done

**Review by arvindkrishnakumar-okta — APPROVED** [review]
lgtm

**Review by arvindkrishnakumar-okta — APPROVED** [review]

File: oauth2/src/main/java/com/okta/spring/boot/oauth/OktaOAuth2AutoConfig.java [hunk 14]
```diff
@@ -70,6 +82,38 @@ OidcClientInitiatedLogoutSuccessHandler oidcLogoutSuccessHandler(OktaOAuth2Prope
...
 
+    @Bean
+    @ConditionalOnMissingBean
+    RestTemplate restTemplate(OktaOAuth2Properties oktaOAuth2Properties) {
+
+        Proxy proxy;
+
+        OktaOAuth2Properties.Proxy proxyProperties = oktaOAuth2Properties.getProxy();
+        Optional<BasicAuthenticationInterceptor> basicAuthenticationInterceptor = Optional.empty();
+        if (proxyProperties != null && Strings.hasText(proxyProperties.getHost()) && proxyProperties.getPort() != 0) {
```
Comment (bdemers) [inline oauth2/src/main/java/com/okta/spring/boot/oauth/OktaOAuth2AutoConfig.java:93]: nit
```suggestion
        if (proxyProperties != null && Strings.hasText(proxyProperties.getHost()) && proxyProperties.getPort() > 0) {
```

File: oauth2/src/main/java/com/okta/spring/boot/oauth/OktaOAuth2ResourceServerAutoConfig.java [hunk 15]
```diff
@@ -64,15 +70,37 @@ JwtDecoder jwtDecoder(OAuth2ResourceServerProperties oAuth2ResourceServerPropert
...
         NimbusJwtDecoder.JwkSetUriJwtDecoderBuilder builder = NimbusJwtDecoder.withJwkSetUri(oAuth2ResourceServerProperties.getJwt().getJwkSetUri());
-        builder.restOperations(restTemplate());
+        builder.restOperations(restTemplate(oktaOAuth2Properties));
         NimbusJwtDecoder decoder = builder.build();
         decoder.setJwtValidator(TokenUtil.jwtValidator(oktaOAuth2Properties.getIssuer(), oktaOAuth2Properties.getAudience()));
         return decoder;
     }
 
-    private RestTemplate restTemplate() {
+    private RestTemplate restTemplate(OktaOAuth2Properties oktaOAuth2Properties) {
```
Comment (bdemers) [inline oauth2/src/main/java/com/okta/spring/boot/oauth/OktaOAuth2ResourceServerAutoConfig.java:79]: Remove the duplicate code from: `OktaOidcUserService`

**Review by bdemers — DISMISSED** [review]
After removing the duplicate code, it looks like it's ready to go

File: oauth2/src/main/java/com/okta/spring/boot/oauth/OktaOAuth2ResourceServerAutoConfig.java (hunk 15 above)
Comment (VitaliiTytarenko-okta) [reply oauth2/src/main/java/com/okta/spring/boot/oauth/OktaOAuth2ResourceServerAutoConfig.java]: Done!

**Pull request merged**